C++ visibility support

jslee02 opened the pull request:
This PR implements the [control of binary symbols visibility](https://gcc.gnu.org/wiki/Visibility) in DART inspired by https://github.com/flexible-collision-library/fcl/pull/233.

`DART_EXPORT` preprocessor is added to any class, struct, and function that needs to be exposed publicly to third-party software. I could compile all the tests, examples, and tutorials without any issue, but it would be good to test DART with other third-party software like AIKIDO for sure.

`DART_HIDE_ALL_SYMBOLS` CMake option is also added, which is `OFF` by default, for the case that we want to hide all the symbols as DART is built as static library.

Caveat:
There is a preprocessor name conflict between the existing `DART_DEPRECATED(version)` and another preprocessor generated by `generate_export_header()`, which is necessary to mark visibility of deprecated APIs. To resolve this, I changed the name to `DART_DEPRECATED2` so that the preprocessor for deprecated API becomes `DART_DEPRECATED2_EXPORT`. I'm open to better workaround.

Alternatively, we could remove `DART_DEPRECATED(version)` and use one generated by `generate_export_header()`. The version information can be provided by using Doxygen comment (e.g., `/// \deprecated Deprecated in <version>. Please use <other_api> instead).`).

This PR is related to #1005.

***

**Before creating a pull request**

- [x] Format new code files using `clang-format`

**Before merging a pull request**

- [x] Set version target by selecting a milestone on the right side
- [x] Summarize this change in `CHANGELOG.md`
- [ ] Add comment to the export preprocessors

mxgrey commented:
This is a good start, but I have a few notes:

> I changed the name to DART_DEPRECATED2 so that the preprocessor for deprecated API becomes DART_DEPRECATED2_EXPORT.

What I've done in the past is:

* Put the auto-generated header into `detail/export.h` (e.g. `dart/detail/export.h`)
* Name the auto-generated macros as `DETAIL_<BASE>_<MACRO_NAME>`
* Manually write an `export.h` file (e.g. `dart/export.h`)
* Have the manually-written `export.h` file *use* the macros that were defined in `dart/detail/export.h`, e.g.:

```
#define DART_EXPORT \
    DETAIL_DART_EXPORT
#define DART_DEPRECATED(v) \
    DETAIL_DART_DEPRECATED
...
```

This also gives us the opportunity to add documentation to each macro, which may be helpful for end users.

I would suggest we consider naming the export macro as `DART_API` instead of the default `DART_EXPORT`, since I think calling it API makes it more explicit about its purpose. We can do that by adding `EXPORT_MACRO_NAME DART_API` to the `generate_export_headers` command.

And a final note: We need to create a unique export header for **each library** in the project, which means each of the optional components that aren't part of `libdart` need their own unique export header with their own unique export macro name.

mxgrey commented:
Revising my last comment: Since we would need a bunch of the "manually written" `export.h` with slight variations, we can create an `export.h.in` and configure it for each of the libraries.

jslee02 commented:
Having another layer of export header sounds good to me.

> And a final note: We need to create a unique export header for each library in the project, which means each of the optional components that aren't part of libdart need their own unique export header with their own unique export macro name.

Could you let me know why do we need different configurations for each component?

mxgrey commented:
The definition of the export macro will be different depending on whether the header is being compiled by the library that is defining the class or by a library that is using the class.

To illustrate how things would go wrong, let's say we use the same export macro for `libdart-gui-osg` as `libdart`. Then when we compile `libdart-gui-osg` and the compiler comes across the declaration of `Skeleton` in `dart/dynamics/Skeleton.hpp`, then the compiler will think that `libdart-gui-osg` wants to define and export the symbol of `dart::dynamics::Skeleton`. This will cause undefined reference and/or other linker errors, since `libdart-gui-osg` does not have access to the definition of `dart::dynamics::Skeleton`.

jslee02 commented:
@mxgrey That makes sense to me. I have a follow-up question though.

Although `gui-osg` component uses different preprocessors (e.g., `DART_GUI_OSG_API`), I think the component still will try to export the symbol of `dart::dynamics::Skeleton` because `Skeleton.hpp` includes `dart/export.hpp` so that `DART_API` is defined to export the symbol. How could we prevent this situation? Should we redefine all the export preprocessor except the preprocessors of the component not to export symbols?

mxgrey commented:
> I think the component still will try to export the symbol of dart::dynamics::Skeleton because Skeleton.hpp includes dart/export.hpp so that DART_API is defined to export the symbol.

This is not correct.

When you use the `generate_export_header` function from cmake, it will add a private compile definition (something along the lines of `dart_EXPORTS`) to the library for which it's generated (this is why the first argument of the function needs to be the library target for which you're generating the export header). That compile definition, which is *only* given to `libdart` during compilation, will make it so that the `DART_API` macro expands to `__declspec(dllexport)` (or whatever the platform-specific export attribute is).

When `libdart-gui-osg` is compiled, it will **not** be given the `dart_EXPORTS` compile definition, so the `DART_API` macro will expand to `__declspec(dllimport)` (or whatever the platform-specific import attribute is).

mxgrey commented:
So the short version is, cmake takes care of your concern automatically.

jslee02 commented:
@mxgrey Thanks for the explanation! I learned a lot while working on this PR.

I've addressed your feedback in the last commits.

`test_Collision` fails due to a reason that is not relevant to the changes of this PR. In a nutshell, the current implementation of `Singleton` doesn't work for multiple binary modules (related [SO post](https://stackoverflow.com/questions/32883872/implementing-singletons-across-binary-modules)). I'll handle this in a future PR.

mxgrey commented on cmake/DARTMacros.cmake:
```diff
@@ -170,3 +170,82 @@ function(dart_format_add)
+  )
+
+  # Generate final export header
+  set(header_guard_name ${base_name_upper}_EXPORT_HPP_)
```
This would probably be much cleaner if we created a `export.hh.in` and then called `configure_file` on it.

jslee02 replied:
Agree. I wrote this lines because I thought each component requires individual `export.hpp.in`, but this is not true.

jslee02 replied:
Agree. I wrote this lines because I thought each component requires individual `export.hpp.in`, but this is not true.

mxgrey commented on cmake/DARTMacros.cmake:
```diff
@@ -170,3 +170,82 @@ function(dart_format_add)
+  generate_export_header(${target_name}
+    EXPORT_MACRO_NAME ${base_name_upper}_DETAIL_API
+    EXPORT_FILE_NAME detail/export.h
+    DEPRECATED_MACRO_NAME ${base_name_upper}_DETAIL_DEPRECATED
```
Nitpick: I like to put the `DETAIL` in front of `DART` so it gets excluded from autocomplete when users start typing `DART...`.

jslee02 replied:
That's true, but I thought we use `DART_` prefix to avoid possible name conflicts. At the same time, I don't see the possibility of name confliction with `DETAIL_DART_~`.

Let me change it to `DETAIL_DART_~` as I don't have a strong preference here.

jslee02 commented:
@mxgrey I've fixed build and test errors. Could you do a second review?

mxgrey commented on dart/common/Deprecated.hpp:
```diff
@@ -42,7 +43,7 @@
+  #define DART_DEPRECATED(version) DETAIL_DART_DEPRECATED
   #define DART_FORCEINLINE __attribute__((always_inline))
 #elif defined(_MSC_VER)
   #define DART_DEPRECATED(version) __declspec(deprecated)
```
This should also be redirected to `DETAIL_DART_DEPRECATED` to be consistent with the other definition.

mxgrey commented:
It looks like the lines that I marked with `Example` didn't make it into the review for some reason. In any case, template definitions shouldn't be given the export macro.

jslee02 commented:
I haven't fully investigated but without the export macro to the template classes caused test failures of `test_Collision` and `test_Skeleton` for different reasons.

`test_Collision` failed due to creating different static instances for collision factory; FCL and DART collision detectors are registered to one factory, but Bullet and FCL are registered to each different factory.

`test_Skeleton` failed due to that null is returned from [here](https://github.com/dartsim/dart/blob/master/unittests/comprehensive/test_Skeleton.cpp#L1280).

mxgrey commented:
> `test_Collision` failed due to creating different static instances for collision factory

I think that's because [this class](https://github.com/dartsim/dart/blob/build/visibility/dart/collision/CollisionDetector.hpp#L65) is never marked as exported. I'm not sure what the correct way to export that would be. Maybe we need to explicitly instantiate the `SingletonFactory` class. In this specific case, we could decide that we do not want users to be able to use our `Singleton` template for any of their own template arguments, but we can't apply that reasoning to all of our templates, or else DART won't be extensible.

> `test_Skeleton` failed due to that null is returned from here.

That might be a trickier issue. It's not immediately clear to me why that would happen.

mxgrey commented:
Yes, I think that would also be dangerous, because the definition is actually given in [the implementation header](https://github.com/dartsim/dart/blob/build/visibility/dart/common/detail/Singleton-impl.hpp#L44).

My understanding of the export macro is that it effectively tells the user's compiler "Don't try to create a definition for this symbol, because the definition will be provided by the library that exported this header", and then it tells the user's linker "Don't look for a definition of this symbol within the library that we just compiled; instead, look for it in one of the libraries that we are linking to".

I think if we "export" that static member, then users will get a linking error for it whenever they pass a custom class type into `Singleton<T>`.
